Add integration for Kostal Piko solar inverters

scheidtdav opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add a new integration for Kostal Piko solar inverters. The api differs dramatically from the already existing kostal_plenticore implementation and requires a different library to access the inverter's data via local poll.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/25002#event-7850279318

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Commit:
Add Kostal Piko inverter integration

Commit:
Use custom device_class for sensor entities

Commit:
Add tests for config flow of kostal_piko

Commit:
Add missing requirements_test_all

Commit:
Add brand file for kostal integrations

Lash-L commented on homeassistant/components/kostal_piko/__init__.py:
```diff
@@ -0,0 +1,118 @@
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+PLATFORMS = ["sensor"]
```
Platform.SENSOR

Lash-L commented on homeassistant/components/kostal_piko/__init__.py:
```diff
@@ -0,0 +1,118 @@
+
+    try:
+        piko = Piko(async_get_clientsession(hass), host, user, password)
+    except Exception as err:
```
Can you seperate invalid auth and other exceptions? It is a lot more helpful for the user that way

scheidtdav replied:
This is done in the config_flow (`async_step_user`), where the connection is tested. When the Piko is first created, there is no actual request sent. So there is also no way of knowing any auth issues, since its active polling.

Lash-L commented on homeassistant/components/kostal_piko/__init__.py:
```diff
@@ -0,0 +1,118 @@
+    coordinator = PikoUpdateCoordinator(hass, piko)
+    await coordinator.async_config_entry_first_refresh()
+
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = coordinator
```
You can combine these

```suggestion
    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = coordinator
```

Lash-L commented on homeassistant/components/kostal_piko/config_flow.py:
```diff
@@ -0,0 +1,79 @@
+SETUP_SCHEMA = vol.Schema(
+    {
+        vol.Required(
+            CONF_HOST, description={"suggested_valed": "http://192.168.178.xyz"}
```
type in value?

Lash-L commented on homeassistant/components/kostal_piko/config_flow.py:
```diff
@@ -0,0 +1,79 @@
+)
+
+
+async def test_connection(hass: HomeAssistant, data) -> tuple[str, str, str]:
```
type data

Lash-L commented on homeassistant/components/kostal_piko/config_flow.py:
```diff
@@ -0,0 +1,79 @@
+
+    async def async_step_user(self, user_input=None) -> FlowResult:
+        """Handle the user initiating the flow via the user interface."""
+        errors = {}
```
type errors

Lash-L commented on homeassistant/components/kostal_piko/config_flow.py:
```diff
@@ -0,0 +1,79 @@
+
+    VERSION = 1
+
+    async def async_step_user(self, user_input=None) -> FlowResult:
```
type user_input

Lash-L commented on homeassistant/components/kostal_piko/manifest.json:
```diff
@@ -0,0 +1,10 @@
+{
+  "domain": "kostal_piko",
+  "name": "Kostal Piko Solar Inverter",
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/kostal_piko",
+  "requirements": ["pykostal==0.0.2"],
+  "codeowners": ["@scheidtdav", "@bedmanforever"],
+  "iot_class": "local_polling",
+  "integration_type": "device"
```
should be in domain, name, alphabetical order

Lash-L commented on homeassistant/components/kostal_piko/translations/de.json:
remove this file - no longer in use

Lash-L commented on homeassistant/components/kostal_piko/translations/en.json:
remove this file - no longer in use

Lash-L commented on homeassistant/components/kostal_piko/translations/sensor.de.json:
remove this file - no longer in use

Lash-L commented on homeassistant/components/kostal_piko/translations/sensor.en.json:
remove this file - no longer in use

Lash-L commented on tests/components/kostal_piko/test_config_flow.py:
```diff
@@ -0,0 +1,172 @@
+from tests.common import MockConfigEntry
+
+
+async def test_show_empty(hass):
```
you need to type hass and have a return type for all of your tests

Lash-L commented on homeassistant/components/kostal_piko/strings.sensor.json:
I don't think this is correct - i'm not 100% though, I'd double check

Lash-L reviewed (changes requested):
Here are my reviews to get your started, I'm not a maintainer, so feel free to ignore anything I said.

Commit:
Add missing types, fix typos

Commit:
Refactor kostal_piko translations

scheidtdav commented:
@Lash-L Thanks for your feedback, much appreciated!
I have implemented all your requested changes and updated to the new translations as well. You were totally right about that.

JrtPec commented on homeassistant/components/kostal_piko/config_flow.py:
```diff
@@ -0,0 +1,88 @@
+) -> tuple[str, str, str]:
+    """Tests the connection to the inverter and returns its name."""
+    session = async_get_clientsession(hass)
+    inverter = kostal.Piko(session, data["host"], data["username"], data["password"])
```
Change the string keys to `CONF_HOST` etc.

JrtPec commented on homeassistant/components/kostal_piko/const.py:
```diff
@@ -0,0 +1,493 @@
+}
+
+
+def round_one(val: int):
```
Did not expect these 2 functions to be in `const.py`

JrtPec commented on homeassistant/components/kostal_piko/sensor.py:
```diff
@@ -0,0 +1,91 @@
+    device_info = DeviceInfo(
+        configuration_url=entry.data[CONF_HOST],
+        identifiers={(DOMAIN, coordinator.data[kostal.InfoVersions.SERIAL_NUMBER])},
+        manufacturer="Kostal",
```
Shouldn't the manufacturer also be a constant?

JrtPec commented on tests/components/kostal_piko/test_config_flow.py:
```diff
@@ -0,0 +1,173 @@
+            },
+        )
+
+        assert result["type"] == "abort"
```
Check against `FlowResultType.ABORT`

JrtPec commented on tests/components/kostal_piko/test_config_flow.py:
```diff
@@ -0,0 +1,173 @@
+        )
+
+        assert "errors" not in result
+        assert result["type"] == "create_entry"
```
Check against `FlowResultType.CREATE_ENTRY`

JrtPec reviewed (commented):
Added a few small comments, mainly related to using constants. Otherwise the code looks nice!

edenhaus commented on homeassistant/components/kostal_piko/sensor.py:
```diff
@@ -0,0 +1,91 @@
+    """Set up the Kostal Piko platform with its sensors."""
+    coordinator = hass.data[DOMAIN][entry.entry_id]
+
+    device_info = DeviceInfo(
+        configuration_url=entry.data[CONF_HOST],
+        identifiers={(DOMAIN, coordinator.data[kostal.InfoVersions.SERIAL_NUMBER])},
+        manufacturer="Kostal",
+        model=coordinator.data[kostal.SettingsGeneral.INVERTER_MAKE],
+        name=coordinator.data[kostal.SettingsGeneral.INVERTER_NAME],
+        sw_version=coordinator.data[kostal.InfoVersions.VERSION_FW],
+        hw_version=coordinator.data[kostal.InfoVersions.VERSION_HW],
+    )
```
```suggestion
```

edenhaus commented on homeassistant/components/kostal_piko/sensor.py:
```diff
@@ -0,0 +1,91 @@
+        super().__init__(coordinator)
+        self.dxs_id = int(description.key)
+        self.formatter = description.formatter
+        self._attr_device_info = deviceInfo
```
```suggestion
        self._attr_device_info = DeviceInfo(
            configuration_url=entry.data[CONF_HOST],
            identifiers={(DOMAIN, coordinator.data[kostal.InfoVersions.SERIAL_NUMBER])},
            manufacturer="Kostal",
            model=coordinator.data[kostal.SettingsGeneral.INVERTER_MAKE],
            name=coordinator.data[kostal.SettingsGeneral.INVERTER_NAME],
            sw_version=coordinator.data[kostal.InfoVersions.VERSION_FW],
            hw_version=coordinator.data[kostal.InfoVersions.VERSION_HW],
        )
```

Makes the code more readable

edenhaus reviewed (changes requested):
I have done an initial review.
Please also address the requested changes of @JrtPec.

scheidtdav commented:
Thanks for your feedback!
I will have a look as soon as possible. I am currently a bit short on time though, so it may take a bit.
